runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,726 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Runtime component for handling disputes of parachain candidates.
+
+use sp_std::prelude::*;
+use sp_std::result;
+#[cfg(feature = "std")]
+use sp_std::marker::PhantomData;
+use primitives::v1::{
+	Id as ParaId, ValidationCode, HeadData, SessionIndex, Hash, BlockNumber, CandidateHash,
+	DisputeState, DisputeStatementSet, MultiDisputeStatementSet, ValidatorId, ValidatorSignature,
+	DisputeStatement, ValidDisputeStatementKind, InvalidDisputeStatementKind,
+	ExplicitDisputeStatement, CompactStatement, SigningContext, ApprovalVote, ValidatorIndex,
+	ConsensusLog,
+};
+use sp_runtime::{
+	traits::{One, Zero, Saturating, AppVerify},
+	DispatchResult, DispatchError, SaturatedConversion,
+};
+use frame_system::ensure_root;
+use frame_support::{
+	decl_storage, decl_module, decl_error, decl_event, ensure,
+	traits::Get,
+	weights::Weight,
+};
+use parity_scale_codec::{Encode, Decode};
+use bitvec::{bitvec, order::Lsb0 as BitOrderLsb0};
+use crate::{
+	configuration::{self, HostConfiguration},
+	initializer::SessionChangeNotification,
+	shared,
+	session_info,
+};
+
+/// Reward hooks for disputes.
+pub trait RewardValidators {
+	// Give each validator a reward, likely small, for participating in the dispute.
+	fn reward_dispute_statement(session: SessionIndex, validators: impl IntoIterator<Item=ValidatorIndex>);
+}
+
+impl RewardValidators for () {
+	fn reward_dispute_statement(_: SessionIndex, _: impl IntoIterator<Item=ValidatorIndex>) { }
+}
+
+/// Punishment hooks for disputes.
+pub trait PunishValidators {
+	/// Punish a series of validators who were for an invalid parablock. This is expected to be a major
+	/// punishment.
+	fn punish_for_invalid(session: SessionIndex, validators: impl IntoIterator<Item=ValidatorIndex>);
+
+	/// Punish a series of validators who were against a valid parablock. This is expected to be a minor
+	/// punishment.
+	fn punish_against_valid(session: SessionIndex, validators: impl IntoIterator<Item=ValidatorIndex>);
```

Comment:
No.  It merely causes escalation, which wastes everyone's time.  I believe the usual slashing system suffices here, including correlation magnification, so enough parallel spam disputes causes 100% slashes, but one lone wrong dispute sounds like 1% or so.  We'd expect the dispute originates from a bug so we'd refund it after forensics.  

Invalid parablocks being accepted otoh violates soundness directly.  We remove all stake supporting an invalid parablocks because our security rests upon gambler's ruin, so immediate 100% slashes that bypass the slash scaling system.